crates/accelerate/src/cos_sin_decomp.rs

```diff
@@ -0,0 +1,215 @@
+        start += 1;
+        end -= 1;
+    }
+}
+
+/// Reverses columns in-place.
+fn reverse_columns(mat: &mut DMatrix<Complex64>) {
+    // Row-wise reverse u
+    let mut start = 0;
+    let mut end = mat.ncols() - 1;
+    while start < end {
+        mat.swap_columns(start, end);
+        start += 1;
+        end -= 1;
+    }
+}
+
+/// Reverses elements in-place.
+fn reverse_vec(vec: &mut DVector<f64>) {
+    let mut start = 0;
+    let mut end = vec.len() - 1;
+    while start < end {
+        vec.swap_rows(start, end);
+        start += 1;
+        end -= 1;
+    }
+}
+
+/// Computes the cosine-sin decomposition (CSD) of a unitary matrix.
+///
+/// # Args
+///
+/// - `u`: The input unitary to decompose
+///
+/// # Returns
+///
+/// The tuple `(l0, l1, r0, r1, theta)` representing the CSD:
+///
+/// ```text
+/// [l0   ] [c -s] [r0   ]
+/// [   l1] [s  c] [   r1]
+/// ```
+///
+/// - `l0`, `l1`, `r0`, and `r1` are unitaries of equal size,
+/// - `theta` is the array of angles, with each angle in in the interval :math:`[0, \pi/2]`.
+///
+/// The diagonal matrices `c` and `s` contain the cosines and sines of angles in `theta`,
+/// respectively. In particular, `c^2 + s^2 = I`.
+///
+/// Furthermore, the angles in `theta` are sorted in descending order, so:
+/// - cosines are in ascending order,
+/// - sines are in descending order.
+pub fn cos_sin_decomposition(u: DMatrix<Complex64>) -> CosSinDecompReturn {
+    let shape = u.shape();
+    let n = shape.0 / 2;
+    // Upper left corner
+    let u00 = u.view((0, 0), (n, n));
+    // Upper right corner
+    let u01 = u.view((0, n), (n, n));
+    // Lower left corner
+    let u10 = u.view((n, 0), (n, n));
+    // Lower right corner
+    let u11 = u.view((n, n), (n, n));
+
+    // Apply SVD to u00
+    let svd = u00.svd(true, true);
+    let mut l0 = svd.u.unwrap();
+    let mut r0 = svd.v_t.unwrap();
+    let mut c: DVector<f64> = svd.singular_values.column(0).into_owned();
+
+    // We have u00 = l0 c r0, where l0 and r0 are unitary, and c is a diagonal matrix
+    // with positive non-decreasing entries. However, we want the entries of c to be
+    // in the ascending order instead. Fortunately, it is easy to modify l0, c, and r0,
+    // so that this becomes true.
+    reverse_rows(&mut r0);
+    reverse_columns(&mut l0);
+    reverse_vec(&mut c);
+
+    // Compute the angles theta. Because u00 is a submatrix of a unitary matrix, its
+    // singular values (the entries of c) are all <= 1. Due to floating-point precision
+    // errors, it is possible that some of the c-values are just a tiny bit larger than 1,
+    // so we correct for that.
+    let thetas: Vec<f64> = c.iter().map(|f| f.min(1.0).acos()).collect();
+
+    // Apply QR to u10_r0_dag.
+    // We have u10 r0* = l1 s, where l1 is unitary and S is upper-triangular.
+    // Equivalently, u10 = l1 s r0.
+    let r0_dag = r0.adjoint();
+    let u10_r0_dag = u10 * r0_dag;
+    let qr = u10_r0_dag.clone().qr();
```

Comment:
I don't think you need this `clone()` the `qr()` method shouldn't need ownership of `u10_r0_dag`:

```suggestion
    let qr = u10_r0_dag.qr();
```

Reply:
Removed.